### ISSUE TITLE
Support intel's SIMD string comparison instructions

### ISSUE THREAD

**Issue by ethernetsellout** [ziglang/zig#16137]
SSE4.2 added the following [string comparison instructions](https://john-lazzaro.github.io/class/special/pdf/string_opcodes.pdf) to the x86 instruction set:
```
pcmpestri - compares explicit length strings, returns index
pcmpistri - compares null terminated strings, returns index
pcmpestri - compares explicit length strings, returns mask
pcmpistri - compares null terminated strings, returns mask
```
These instructions are quite versatile, offering multiple modes of comparison determined by the imm8 operand. They are particularly useful for parsing. LLVM has intrinsics for them.

Potential builtin: 
```zig
@cmpstr(haystack, needle, comptime compare_mode: Mode) ReturnType
```
```zig
const Mode = enum { any, each, ordered, range_gt, range_lt };
```
Where haystack & needle are a vector of u8s, and the result type is an integer or a vector of bools.

Implementing the builtin when the CPU does not have said instructions could be a challenge. Nonetheless it would be nice to see these instructions supported.

**Comment by Validark** [comment]
I am quite interested in these instructions and have experimented with them. I am not sure in how many cases they actually are a net win over alternative vector instructions which can perform the same task. I used `pcmpistri` (the fastest one?) to implement a function which returns the common prefix length between two strings and the performance was worse for my use-case. However, I am on a Zen 2 architecture, so it's possible more recent architectures optimize it more than my machine does. It's also possible that the non "equal each" modes are more beneficial for their use-cases.

---

I do not think these are a good fit for a language builtin. The reason I say this is because `@cmpstr` is not a single semantic unit. It's really a collection of functions built-in to x86_64 hardware. We call it a single "instruction" but the different modes make it do ENTIRELY different things. Regardless of what the hardware implementation looks like, they could have defined each of the modes as separate instructions, and in code, I would move them into separate functions. If there was any other built-in function in the language or standard library with a function signature that did entirely different and semantically unrelated things based on a flag, it would rightfully be considered a terrible API in my opinion.

If you want any specific vector semantics to be given more support, they should each be considered separately. (Each "mode" should have its own proposal.) One thing I notice is that doing an "equal each" can already be done with `vec1 == vec2` in Zig, so supporting that mode in `@cpmstr` does not really add anything semantically to the language that wasn't already trivial with existing language support. The "equal any" and "range" operations seem nice but I think each of those should get their own functions. I also am not sure they qualify as fundamental vector operations worthy of a builtin. Perhaps a proposal would change my mind but these operations seem more like standard library functions to me.

tldr; I do think Zig needs to support intrinsics and this should be one of them. However, as a single semantic unit it supports too many entirely separate things for one function and I do not consider the operations it supports to be fundamental vector operations. I think these feel more like standard library functions.

[Here's some inline assembly code that uses pcmpistri, if you are interested.](https://github.com/Validark/DynSDT/blob/fbefb0f592ed6f821b41b23ed90c416541c10c05/zig/src/main.zig#L242-L272)
